internal/consul/namespaces.go

```diff
@@ -45,3 +66,76 @@ func EnsureNamespaceExists(client Client, ns string) (bool, error) {
 	}
 	return true, err
 }
+
+func getOrCreateCrossNamespacePolicy(client Client, partitionInfo PartitionInfo) (*api.ACLPolicy, error) {
+	acl := client.ACL()
+
+	rules, err := crossNamespaceRules(partitionInfo)
+	if err != nil {
+		return &api.ACLPolicy{}, err
+	}
+	policy := &api.ACLPolicy{
+		Name:        "cross-namespace-policy",
+		Description: "Policy to allow permissions to cross Consul namespaces for k8s services",
+		Rules:       rules,
+	}
+	createdPolicy, _, err := acl.PolicyCreate(policy, nil)
+	if err != nil && !isPolicyExistsErr(err, policy.Name) {
+		return nil, err
+	}
+
+	// this means the policy is newly created and we can just return it
+	if createdPolicy != nil {
+		return createdPolicy, nil
+	}
+
+	// here the policy already exists so we need to fetch it
+	createdPolicy, _, err = acl.PolicyReadByName(policy.Name, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if createdPolicy == nil {
+		return nil, errors.New("failed to read policy \"cross-namespace-policy\" from consul server")
+	}
+	return createdPolicy, nil
+}
+
+func crossNamespaceRules(partitionInfo PartitionInfo) (string, error) {
+	crossNamespaceRulesTpl := `{{- if .EnablePartitions }}
+partition "{{ .PartitionName }}" {
+{{- end }}
+  namespace_prefix "" {
+    service_prefix "" {
+      policy = "read"
+    }
+    node_prefix "" {
+      policy = "read"
+    }
+  }
+{{- if .EnablePartitions }}
+}
+{{- end }}`
+
+	compiled, err := template.New("root").Parse(strings.TrimSpace(crossNamespaceRulesTpl))
+	if err != nil {
+		return "", err
+	}
+
+	// Render the template
+	var buf bytes.Buffer
+	err = compiled.Execute(&buf, partitionInfo)
+	if err != nil {
+		// Discard possible partial results on error return
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// isPolicyExistsErr returns true if err is due to trying to call the
+// policy create API when the policy already exists.
+func isPolicyExistsErr(err error, policyName string) bool {
+	return strings.Contains(err.Error(), "Unexpected response code: 500") &&
```

Comment:
Can you add a quick comment here that this handles the case of ACLs not being enabled? It's always confusing that the only way to do this is to issue something against the policy endpoint and see if Consul blows up.

Reply:
done and done